### PR TITLE
 Remove bindings of removed paths from scope

### LINKED ISSUE #6923: TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)
Choose one: bug report

The error message

> TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)

is shown when transpiling the input code. node successfully executes the input code.

### Input Code

```js
async function foo() {
	(async function (number) {
		const tmp = number
	})
}
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
{ "presets": [ [ "@babel/env"
               , { "targets": { "browsers": [ "last 2 chrome versions"
                                            , "last 2 chromeandroid versions"
                                            , "firefox esr"
                                            , "last 2 firefox versions"
                                            , "edge >= 15"
                                            , "safari >= 11"
                                            ]
                              }
                 , "debug": true
                 }
               ]
             ]
}
```

```js
{
  "dependencies": {
    "@babel/cli": "^7.0.0-beta.32",
    "@babel/core": "^7.0.0-beta.32",
    "@babel/preset-env": "^7.0.0-beta.32",
    "regenerator-runtime": "^0.11.0"
  }
}
```

### Expected Behavior

The script is successfully transpiled.

### Current Behavior

The given error message appears:

```
root@3f46f465fd04:/pwd# yarn run babel test.js
yarn run v1.3.2
warning package.json: No license field
$ /pwd/node_modules/.bin/babel test.js
@babel/preset-env: `DEBUG` option

Using targets:
{
  "chrome": "61",
  "edge": "15",
  "firefox": "52",
  "safari": "11"
}

Using modules transform: commonjs

Using plugins:
  transform-destructuring { "edge":"15", "firefox":"52" }
  transform-for-of { "firefox":"52" }
  transform-function-name { "edge":"15", "firefox":"52" }
  transform-literals { "firefox":"52" }
  transform-parameters { "firefox":"52" }
  transform-regenerator { "firefox":"52" }

Using polyfills: No polyfills were added, since the `useBuiltIns` option was not set.
TypeError: test.js: Duplicate declaration "tmp" (This is an error on an internal node. Probably an internal error.)
    at File.buildCodeFrameError (/pwd/node_modules/@babel/core/lib/transformation/file/file.js:209:12)
    at Scope.checkBlockScopedCollisions (/pwd/node_modules/@babel/traverse/lib/scope/index.js:295:27)
    at Scope.registerBinding (/pwd/node_modules/@babel/traverse/lib/scope/index.js:477:16)
    at Scope.registerDeclaration (/pwd/node_modules/@babel/traverse/lib/scope/index.js:396:14)
    at Object.Declaration (/pwd/node_modules/@babel/traverse/lib/scope/index.js:89:12)
    at NodePath._call (/pwd/node_modules/@babel/traverse/lib/path/context.js:64:19)
    at NodePath.call (/pwd/node_modules/@babel/traverse/lib/path/context.js:38:17)
    at NodePath.visit (/pwd/node_modules/@babel/traverse/lib/path/context.js:95:12)
    at TraversalContext.visitQueue (/pwd/node_modules/@babel/traverse/lib/context.js:139:18)
    at TraversalContext.visitMultiple (/pwd/node_modules/@babel/traverse/lib/context.js:93:17)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

### Context

It worked fine with `babel-preset-env` in version `1.5.2`. It is broken with `@babel/preset-env` in version `7.0.0-beta.32`.

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.32 (@babel/core 7.0.0-beta.32)
| Babylon          | 
| node             | v9.2.0
| npm              | 5.5.1
| Operating System | Official node.js Docker container

### LINKED ISSUE #7304: [preset-env] Duplicate declaration "object"
### Is this a bug report or feature request?

Bug

[Repro repo](https://github.com/julien-f/issue-babel-env-duplicate).

### Input Code

```js
export default class {
  method ({ ...object }) {}
}
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
{
  "presets": [
    [
      "@babel/env",
      {
        "shippedProposals": true,
        "targets": {
          "node": 8
        },
        "useBuiltIns": "usage"
      }
    ]
  ]
}
```

### Expected Behavior

Should compile.

### Current Behavior

```
TypeError: index.js: Duplicate declaration "object"
  1 | export default class {
> 2 |   method ({ ...object }) {}
    |                ^
  3 | }
  4 |
    at File.buildCodeFrameError (issue-babel-env-duplicate/node_modules/@babel/core/lib/transformation/file/file.js:209:12)
    at Scope.checkBlockScopedCollisions (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/scope/index.js:303:27)
    at Scope.registerBinding (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/scope/index.js:485:16)
    at Scope.registerDeclaration (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/scope/index.js:404:14)
    at Object.BlockScoped (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/scope/index.js:146:28)
    at Object.newFn (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/visitors.js:266:17)
    at NodePath._call (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/path/context.js:64:19)
    at NodePath.call (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/path/context.js:34:14)
    at NodePath.visit (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/path/context.js:99:12)
    at TraversalContext.visitQueue (issue-babel-env-duplicate/node_modules/@babel/traverse/lib/context.js:139:18)
```

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.39
| Babylon          | 
| node             | 9.4.0
| npm              | 
| Operating System |

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #7304, fixes #6923 
| Patch: Bug Fix?          | :+1: 
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

### CODE REVIEW

**Comment by benjamn** [comment]
Awesome, glad to know my [hunch](https://github.com/babel/babel/issues/7304#issuecomment-362034960) was accurate! Thanks for working on this.

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/output.js [hunk 1]
```diff
@@ -10,10 +10,11 @@ var {
 } = z,
     y1 = babelHelpers.objectWithoutProperties(z, ["x1"]);
 x1++;
-var {
-  [a]: b
+var _a = a,
```
Comment (Andarist) [inline packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/output.js]: is it necessary to split this computed property like this?

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+it("es7.objectRestSpread", () => {
```
Comment (Andarist) [inline packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js]: hm, is this some kind of hidden feature of the test runner? or just a sample input that accidentally looks like a test? :)

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
+"use strict";
+
+Object.defineProperty(exports, "__esModule", {
+  value: true
+});
+exports.default = void 0;
+
+class _default {
```
Comment (Andarist) [inline packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js]: This output seems to be buggy regarding what we discuss (cyclic deps and all) under babel-helper-split-export-declaration PR, right?

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js (hunk 2 above)
Comment (benjamn) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js]: It's the `input.js` from the test I added, left over from reducing a real test case. The extra scope is necessary, though the exact string literal doesn't matter.

File: packages/babel-traverse/src/path/removal.js [hunk 4]
```diff
@@ -17,6 +18,13 @@ export function remove() {
   this._markRemoved();
 }
 
+export function _removeFromScope() {
+  const bindings = this.getBindingIdentifiers();
+  for (const name of Object.keys(bindings)) {
```
Comment (Andarist) [inline packages/babel-traverse/src/path/removal.js]: wouldnt a simple forEach be more readable here?

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js (hunk 2 above)
Comment (Andarist) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/duplicate-decl-bug/input.js]: i was wondering about `it`, good to know that it's "just" a sample input 👍

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/output.js (hunk 1 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/object-rest/variable-destructuring/output.js]: It's a bug. I'm looking at it right now

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js (hunk 3 above)
Comment (nicolo-ribaudo) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js]: I don't think that class declarations are hoisted? :thinking: 
```js
new A(); // ReferenceError
class A {}:
```

Anyway, that code is not related to this PR, it is generated either by transform-classes or transform-modules-commonjs.

**Comment by nodkz** [comment]
@nicolo-ribaudo can you test following case:
```js
const obj = { a: 1, b: 2, c: 3 };
const { a, ...rest } = obj;
const newObj = { ...{ d: 4 }, ...rest };
```

I got error `Duplicate declaration "rest"`

Thanks.

**Comment by nicolo-ribaudo** [comment]
It looks like it works. You can test this PR at https://babeljs.io/repl/build/6813/

File: packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js (hunk 3 above)
Comment (Andarist) [reply packages/babel-plugin-proposal-object-rest-spread/test/fixtures/regression/gh-7304/output.js]: after experimenting with modules this output seems correct anyway

**Comment by nicolo-ribaudo** [comment]
I'm merging this PR.
Note that currently it isn't possible to have the Scope 100% synchronized with the path (since it is mutable), so problems like this might come up in totally unrelated parts of our codebase.

**Comment by nicolo-ribaudo** [comment]
The failing test is not related (https://github.com/babel/babel/pull/7344)

**Pull request merged**